storage/src/adb/any/mod.rs

```diff
@@ -198,48 +200,74 @@ impl<E: RStorage + Clock + Metrics, K: Array, V: Array, H: CHasher, T: Translato
         Ok(db)
     }
 
-    /// Initialize an [Any] database in a pruned state.
-    /// Removes all existing persisted data and applies the state given in `cfg`.
+    /// Returns an [Any] initialized from sync data in `cfg` for use in state sync.
+    ///
+    /// # Behavior
+    ///
+    /// This method handles different initialization scenarios based on existing data:
+    /// - If the MMR journal is empty or the last item is before `lower_bound`, it creates a
+    ///   fresh MMR from the provided `pinned_nodes`
+    /// - If the MMR journal has data but is incomplete (< `upper_bound`), missing operations
+    ///   from the log are applied to bring it up to the target state
+    /// - If the MMR journal has data beyond the `upper_bound`, it is rewound to match the sync target
+    ///
+    /// # Returns
+    ///
+    /// An [Any] db populated with the state from `cfg.lower_bound` to `cfg.upper_bound`, inclusive.
+    /// The pruning boundary is set to `cfg.lower_bound`.
     pub(super) async fn init_synced(
         context: E,
         cfg: SyncConfig<E, K, V, T, H::Digest>,
     ) -> Result<Self, Error> {
-        let mmr_config = crate::mmr::journaled::SyncConfig {
-            config: MmrConfig {
-                journal_partition: cfg.db_config.mmr_journal_partition,
-                metadata_partition: cfg.db_config.mmr_metadata_partition,
-                items_per_blob: cfg.db_config.mmr_items_per_blob,
-                write_buffer: cfg.db_config.mmr_write_buffer,
-                thread_pool: cfg.db_config.thread_pool.clone(),
-                buffer_pool: cfg.db_config.buffer_pool,
+        let mut mmr = Mmr::init_sync(
+            context.with_label("mmr"),
+            crate::mmr::journaled::SyncConfig {
+                config: crate::mmr::journaled::Config {
+                    journal_partition: cfg.db_config.mmr_journal_partition,
+                    metadata_partition: cfg.db_config.mmr_metadata_partition,
+                    items_per_blob: cfg.db_config.mmr_items_per_blob,
+                    write_buffer: cfg.db_config.mmr_write_buffer,
+                    thread_pool: cfg.db_config.thread_pool.clone(),
+                    buffer_pool: cfg.db_config.buffer_pool.clone(),
+                },
+                lower_bound: leaf_num_to_pos(cfg.lower_bound),
+                // The last node of an MMR with `cfg.upper_bound` + 1 operations is at the position
+                // right before where the next leaf goes.
+                upper_bound: leaf_num_to_pos(cfg.upper_bound + 1) - 1,
```

Comment:
I suppose this could panic from underflow (and/or upper_bound may be less than lower_bound).

We don't validate arguments like this elsewhere in storage (yet) but calling out in case there is anything you think we should check for some reason.